### PR TITLE
new signup flow added

### PR DESCRIPTION
**Pull request by ivinayakg**
### Issue
 how [#688](https://github.com/Real-Dev-Squad/website-backend/issues/688) will behave on the front end.
 [backend PR](https://github.com/Real-Dev-Squad/website-backend/pull/694), the code for this feature.

### What is the change?
- the webpage from which the request is initiated, its href (URL) will be passed to the request as a query param and later after authentication backend will make sure to redirect our user back to the webpage

### Is Development Tested?
- [ ] Yes
- [x] No

### CODE REVIEW

File: js/github.js [hunk 1]
```diff
@@ -9,10 +10,38 @@ const doesGitHubCookieExist = () => {
   return !!githubCookieStr;
 };
 
+const SIGN_IN_GITHUB_OAUTH = () => {
+  let originURL = window?.location?.origin ?? '/';
```
Comment (rohan09-raj) [inline js/github.js:14]: Is optional chaining necessary ??

File: js/github.js [hunk 2]
```diff
@@ -9,10 +10,38 @@ const doesGitHubCookieExist = () => {
...
+  let githubOauth =
+    'https://github.com/login/oauth/authorize?client_id=23c78f66ab7964e5ef97&state=';
```
Comment (rohan09-raj) [inline js/github.js:16]: Can we please make it a constant ?

File: js/github.js [hunk 3]
```diff
@@ -9,10 +10,38 @@ const doesGitHubCookieExist = () => {
...
+  let encodedOriginURL = btoa(originURL);
+
+  return githubOauth + encodedOriginURL;
+};
+
 const updateGitHubLink = () => {
+  const githubMock = JSON.parse(localStorage.getItem('githubMock')) ?? {};
```
Comment (rohan09-raj) [inline js/github.js:23]: Can we name it better ?? What actually it is, a URL or something else ?

File: js/github.js [hunk 4]
```diff
@@ -9,10 +10,38 @@ const doesGitHubCookieExist = () => {
   return !!githubCookieStr;
 };
 
+const SIGN_IN_GITHUB_OAUTH = () => {
```
Comment (rohan09-raj) [inline js/github.js:13]: Can we please use CamelCase here?

File: js/github.js [hunk 5]
```diff
@@ -9,10 +10,38 @@ const doesGitHubCookieExist = () => {
...
+  }
+
   const allLoginBtns = document.querySelectorAll('.btn-login');
   allLoginBtns.forEach((btn) => {
-    btn.setAttribute('href', 'https://github.realdevsquad.com');
+    btn.setAttribute('href', signUpLink);
+
+    btn.addEventListener('click', (event) => {
+      event.preventDefault();
+      let githubMock = {
```
Comment (rohan09-raj) [inline js/github.js:38]: Can it be made `const` ?

File: js/github.js [hunk 6]
```diff
@@ -9,10 +10,38 @@ const doesGitHubCookieExist = () => {
...
+        visited: true,
+        expiresIn: new Date().getTime() + 15 ** 9,
+      };
+      localStorage.setItem('githubMock', JSON.stringify(githubMock));
```
Comment (rohan09-raj) [inline js/github.js:42]: Can we please make the key a constant, as it is used at many places ?

File: js/github.js (hunk 2 above)
Comment (ivinayakg) [reply js/github.js:16]: sure noted

File: js/github.js [hunk 7]
```diff
@@ -9,10 +9,36 @@ const doesGitHubCookieExist = () => {
   return !!githubCookieStr;
```
Comment (RitikJaiswal75) [inline js/github.js:9]: Where are you getting githubCookieStr from?

File: js/github.js (hunk 5 above)
Comment (RitikJaiswal75) [reply js/github.js:35]: are we planning to export it?

Comment (rohan09-raj) [reply js/github.js:35]: Nope I just wanted the variable to be a constant, so that it can't be reassigned

File: js/github.js (hunk 7 above)
Comment (rohan09-raj) [inline js/github.js:9]: What is the point of adding two `!!` not operators ?? Can we please correct it to only `githubCookieStr` ?

File: js/github.js (hunk 1 above)
Comment (rohan09-raj) [reply js/github.js:13]: You have already used the null coalescing operator, so it won't throw any error if the `window.location.origin` is undefined or null

File: js/github.js [hunk 8]
```diff
@@ -9,10 +9,36 @@ const doesGitHubCookieExist = () => {
...
 };
 
+const signInGithubOAuth = () => {
+  let originURL = window?.location?.origin ?? '/';
+  let encodedOriginURL = btoa(originURL);
```
Comment (rohan09-raj) [inline js/github.js:14]: Can we please make this variables as constants ?

File: js/github.js [hunk 9]
```diff
@@ -9,10 +14,36 @@ const doesGitHubCookieExist = () => {
   return !!githubCookieStr;
 };
 
+const signInGithubOAuth = () => {
```
Comment (rohan09-raj) [inline js/github.js:17]: Typo - `Github` -> `GitHub`

File: js/github.js [hunk 10]
```diff
@@ -9,10 +14,36 @@ const doesGitHubCookieExist = () => {
...
+};
+
 const updateGitHubLink = () => {
+  const githubMock = JSON.parse(localStorage.getItem(GITHUB_OBJECT_KEY)) ?? {};
+  const isGithubMockExpired =
+    new Date().getTime() > (githubMock.expiresIn ?? 0);
+
+  let signUpLink = GITHUB_MOCK_URL;
+  if (githubMock.visited && !isGithubMockExpired) {
+    signUpLink = signInGithubOAuth();
```
Comment (rohan09-raj) [inline js/github.js:30]: Please update the function name on Line 30 where you have called it `signInGitHubOAuth` and verify if changes are working fine or not

**Review by prakashchoudhary07 — APPROVED** [review]
Thanks for the fixes

**Pull request merged**